File: vqa-rnn-mlp-v2/session_manager.lua
--Provides some ability to manage runs
--  Creates a separate folder for each run
--  Store the parameters as a json file so it can be retrieved later
--  It is a work-in-progress partial implementation but useful

local cjson=require('cjson');
local torch=require('torch');
local paths=require('paths');
local dirname_length=12;
local Session={};
local function readAll(file)
    local f = io.open(file, "r")
	if f==nil then
		error({msg='Failed to open file',file=file});
	end
    local content = f:read("*all");
    f:close()
    return content;
end
local function loadJson(fname)
	local t=readAll(fname);
	return cjson.decode(t);
end
local function writeAll(file,data)
    local f = io.open(file, "w");
	if f==nil then
		error({msg='Failed to open file',file=file});
	end
    f:write(data);
    f:close();
end
local function saveJson(fname,t)
	return writeAll(fname,cjson.encode(t));
end
--hash a table of strings, numbers and nil
local function hash(t)
	return 0;
end
local function random_string(l,prefix)
	prefix=prefix or '';
	local s=prefix;
	local candidates='ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	for i=1,l do
		local ind=math.random(string.len(candidates));
		s=s..candidates:sub(ind,ind);
	end
	return s;
end
function Session:init(root)
	local session={};
	setmetatable(session, self);
    self.__index = self;
	session.root=root;
	return session;
end
--Create a new session
--Params: a table of parameters for this session. Only supports parameters that can be stored as json.
function Session:new(params)
	local dirname='';
	--Create a random directory
	while true do
		dirname=random_string(dirname_length,'session_');
		if paths.mkdir(paths.concat(self.root,dirname)) then
			break;
		end
	end
	--Write the parameters into the directory
	saveJson(paths.concat(self.root,dirname,'_session_config.json'),params);
	--Return the path to the directory
	return paths.concat(self.root,dirname);
end
--Find the dirname for a particular session. TODO
function Session:find(params)
	return paths.concat(self.root,'');
end
function Session:list()
	local dirs={};
	local params={};
	for dir in paths.iterdirs(self.root) do
		local param={};
		local root=self.root;
		if pcall(function () param=loadJson(paths.concat(root,dir,'_session_config.json')) end) then
			table.insert(dirs,dir);
			table.insert(params,param);
		else
			print(string.format('Warning: Could not find session in %s',dir));
		end
	end
	return {dirs=dirs,params=params};
end

--Load the configuration for a particular session by name.
function Session:get_params(session_name)
	return loadJson(paths.concat(self.root,session_name,'_session_config.json'));
end
return Session;

File: mnist-mlp/util/session_manager.py
import os
import os.path
import sys

class Session:
	id=-1;
	def __init__(self,id=-1):
		if not os.path.isdir('sessions'):
			os.mkdir('sessions');
		if id==-1:
			#New session
			existing_sessions=dict();
			for dir in os.listdir('sessions'):
				existing_sessions[dir]=1;
			id=0;
			while True:
				if '%07d'%id in existing_sessions:
					id=id+1;
				else:
					break;
			os.mkdir(os.path.join('sessions','%07d'%id));
		elif not os.path.isdir(os.path.join('sessions','%07d'%id)):
			os.mkdir(os.path.join('sessions','%07d'%id));
		self.id=id;
		return;
	#
	def file(self,p1='',p2='',p3=''):
		if not os.path.isdir(os.path.join('sessions','%07d'%self.id)):
			os.mkdir(os.path.join('sessions','%07d'%self.id));
		if p1=='':
			return os.path.join('sessions','%07d'%self.id);
		elif p2=='':
			return os.path.join('sessions','%07d'%self.id,p1);
		elif p3=='':
			if not os.path.isdir(os.path.join('sessions','%07d'%self.id,p1)):
				os.mkdir(os.path.join('sessions','%07d'%self.id,p1));
			return os.path.join('sessions','%07d'%self.id,p1,p2);
		else:
			if not os.path.isdir(os.path.join('sessions','%07d'%self.id,p1)):
				os.mkdir(os.path.join('sessions','%07d'%self.id,p1));
			if not os.path.isdir(os.path.join('sessions','%07d'%self.id,p1,p2)):
				os.mkdir(os.path.join('sessions','%07d'%self.id,p1,p2));
			return os.path.join('sessions','%07d'%self.id,p1,p2,p3);
	
	def log(self,str,fname='log.txt'):
		print(str);
		sys.stdout.flush()
		f=open(self.file(fname),'a');
		f.write(str+'\n');
		f.close();
	
	def log_test(self,str):
		print(str);
		sys.stdout.flush()
		f=open(self.file('test.txt'),'a');
		f.write(str+'\n');
		f.close();
